zerver/actions/presence.py

```diff
@@ -77,39 +81,73 @@ def do_update_user_presence(
 ) -> None:
     client = consolidate_client(client)
 
+    # TODO: While we probably DO want creating an account to
+    # automatically create a first `UserPresence` object with
+    # last_connected_time and last_active_time as the current time,
+    # our presence tests don't understand this, and it'd be perhaps
+    # wrong for some cases of account creation via the API.  So we may
+    # want a "never" value here as the default.
     defaults = dict(
-        timestamp=log_time,
-        status=status,
+        # Given that these are defaults for creation of a UserPresence row
+        # if one doesn't yet exist, the most sensible way to do this
+        # is to set both last_active_time and last_connected_time
+        # to log_time.
+        last_active_time=log_time,
+        last_connected_time=log_time,
         realm_id=user_profile.realm_id,
     )
+    if status == UserPresence.LEGACY_STATUS_IDLE_INT:
+        # If the presence entry for the user is just to be created, and
+        # we want it to be created as idle, then there needs to be an appropriate
+        # offset between last_active_time and last_connected_time, since that's
+        # what the event-sending system calculates the status based on, via
+        # format_legacy_presence_dict.
+        defaults["last_active_time"] = log_time - datetime.timedelta(
+            seconds=settings.PRESENCE_LEGACY_EVENT_OFFSET_FOR_ACTIVITY_SECONDS + 1
+        )
```

Comment:
Hmm, I wonder if it'd be better to address the TODO above by making `None` be a supported value for both `last_connected_time` and `last_active_time`? That would affect the future API, but it might be appropriate for both the case in this block (we'd set `last_active_time=None`, but `last_connected_time` would be set as `log_time`) and also the "New account created" block.

Would require some adjustments.

Reply:
@timabbott Yeah we should add that option though I'm thinking this should probably be done as an iteration on this after we test the current state of this on czo? This will add some more complexity, so that might make it easier to review and identify the source of bugs if we run into any on czo?